# Mutual Fund Plan

##### Mutual funds are investment plans that pool money from multiple investors to purchase a diversified portfolio of stocks, bonds, and other securities, managed by professional fund managers. A mutual fund plan is created by selecting the stocks where an investor can benefit in the long term. 

A mutual fund plan is created by selecting the stocks where an investor can benefit in the long term. we can follow below process to create a mutual fund plan:

- Step 1. Gather historical stock data, such as closing prices and growth trends over time.
- Step 2: Calculate key metrics like Return on Investment (ROI) and volatility (risk) to understand how each stock has performed historically.
- Step 3: Choose stocks that have a high ROI and low volatility to ensure a balance between risk and reward.
- Step 4: Calculate the future value of monthly investments based on the expected ROI of the selected stocks.

In [1]:
# Importing requried libraries
import pandas as pd
import numpy as np
import plotly.express as px
import plotly.graph_objs as go
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df = pd.read_csv('nifty50_closing_prices.csv')
df.head()

Date  RELIANCE.NS  HDFCBANK.NS  ICICIBANK.NS  \
0  2024-08-20 00:00:00+05:30  2991.899902  1637.699951   1179.449951   
1  2024-08-21 00:00:00+05:30  2997.350098  1625.800049   1174.849976   
2  2024-08-22 00:00:00+05:30  2996.250000  1631.300049   1191.099976   
3  2024-08-23 00:00:00+05:30  2999.949951  1625.050049   1203.500000   
4  2024-08-26 00:00:00+05:30  3025.199951  1639.949951   1213.300049   

       INFY.NS       TCS.NS  KOTAKBANK.NS  HINDUNILVR.NS      ITC.NS  \
0  1872.199951  4523.299805   1805.650024    2751.050049  498.799988   
1  1872.699951  4551.500000   1812.949951    2791.199951  505.399994   
2  1880.250000  4502.000000   1821.500000    2792.800049  504.549988   
3  1862.099976  4463.899902   1818.000000    2815.600098  505.799988   
4  1876.150024  4502.450195   1812.500000    2821.149902  505.700012   

         LT.NS  ...  HEROMOTOCO.NS   DRREDDY.NS   SHREECEM.NS  BRITANNIA.NS  \
0  3572.699951  ...    5244.399902  6965.350098  24730.550781   5765.799805   
1  3596.050049  ...    5284.700195  7062.450195  24808.050781   5837.350098   
2  3606.500000  ...    5329.950195  6969.049805  25012.400391   5836.799805   
3  3598.550049  ...    5384.899902  6954.500000  24706.050781   5792.649902   
4  3641.899902  ...    5343.750000  6943.299805  24906.449219   5796.950195   

       UPL.NS  EICHERMOT.NS   SBILIFE.NS  ADANIPORTS.NS  BAJAJ-AUTO.NS  \
0  566.150024   4883.250000  1761.300049    1492.550049    9779.700195   
1  568.299988   4913.549805  1800.599976    1503.500000    9852.000000   
2  579.150024   4933.549805  1795.250000    1492.300049    9914.200195   
3  573.700012   4898.100098  1789.300049    1491.300049   10406.450195   
4  577.450012   4875.200195  1796.250000    1482.550049   10432.549805   

   HINDALCO.NS  
0   672.900024  
1   685.599976  
2   685.549988  
3   685.099976  
4   711.849976  

[5 rows x 51 columns]

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 24 entries, 0 to 23
Data columns (total 51 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Date           24 non-null     object 
 1   RELIANCE.NS    24 non-null     float64
 2   HDFCBANK.NS    24 non-null     float64
 3   ICICIBANK.NS   24 non-null     float64
 4   INFY.NS        24 non-null     float64
 5   TCS.NS         24 non-null     float64
 6   KOTAKBANK.NS   24 non-null     float64
 7   HINDUNILVR.NS  24 non-null     float64
 8   ITC.NS         24 non-null     float64
 9   LT.NS          24 non-null     float64
 10  SBIN.NS        24 non-null     float64
 11  BAJFINANCE.NS  24 non-null     float64
 12  BHARTIARTL.NS  24 non-null     float64
 13  HCLTECH.NS     24 non-null     float64
 14  ASIANPAINT.NS  24 non-null     float64
 15  AXISBANK.NS    24 non-null     float64
 16  DMART.NS       24 non-null     float64
 17  MARUTI.NS      24 non-null     float64
 18  ULTRACEMCO.N

In [4]:
df.describe()

RELIANCE.NS  HDFCBANK.NS  ICICIBANK.NS      INFY.NS       TCS.NS  \
count    24.000000    24.000000     24.000000    24.000000    24.000000   
mean   2976.912506  1652.339579   1236.770818  1914.558324  4478.349976   
std      41.290551    28.258220     36.438726    30.240685    70.822718   
min    2903.000000  1625.050049   1174.849976  1862.099976  4284.899902   
25%    2941.862488  1637.062469   1219.750031  1893.687469  4472.937622   
50%    2988.924927  1640.875000   1235.474976  1911.225037  4504.050049   
75%    3005.237427  1666.112457   1250.799988  1941.862488  4514.362549   
max    3041.850098  1741.199951   1338.449951  1964.500000  4553.750000   

       KOTAKBANK.NS  HINDUNILVR.NS      ITC.NS        LT.NS     SBIN.NS  ...  \
count     24.000000      24.000000   24.000000    24.000000   24.000000  ...   
mean    1809.422918    2845.333344  507.739581  3647.099976  802.233332  ...   
std       32.936318      65.620694    5.472559    60.511574   17.442330  ...   
min     1764.150024    2751.050049  497.299988  3536.949951  768.599976  ...   
25%     1783.800049    2790.662476  504.962502  3597.950012  785.224991  ...   
50%     1804.500000    2838.699951  507.550003  3646.349976  811.950012  ...   
75%     1822.987488    2899.987549  511.337509  3689.325012  816.050018  ...   
max     1904.500000    2977.600098  519.500000  3793.899902  824.799988  ...   

       HEROMOTOCO.NS   DRREDDY.NS   SHREECEM.NS  BRITANNIA.NS      UPL.NS  \
count      24.000000    24.000000     24.000000     24.000000   24.000000   
mean     5619.377096  6785.795817  25299.906169   5935.202026  596.343750   
std       247.092728   175.124908    429.919834    144.164343   16.975821   
min      5244.399902  6502.549805  24692.199219   5703.350098  566.150024   
25%      5370.012451  6649.574951  24885.174316   5835.449829  578.875015   
50%      5662.075195  6741.475098  25432.200195   5919.099854  600.850006   
75%      5783.537598  6956.612549  25630.112793   6074.562500  610.737518   
max      6013.250000  7062.450195  26019.650391   6210.549805  619.200012   

       EICHERMOT.NS   SBILIFE.NS  ADANIPORTS.NS  BAJAJ-AUTO.NS  HINDALCO.NS  
count     24.000000    24.000000      24.000000      24.000000    24.000000  
mean    4863.831258  1849.331243    1462.916677   10999.654134   681.885417  
std       68.442418    43.189734      26.223794     659.810841    15.952804  
min     4726.649902  1761.300049    1408.199951    9779.700195   647.700012  
25%     4846.037598  1820.725006    1440.199951   10617.962402   672.162521  
50%     4871.925049  1845.099976    1470.950012   10927.625000   684.050018  
75%     4912.350098  1879.149963    1482.062531   11697.137207   687.799988  
max     4963.149902  1928.650024    1503.500000   11950.299805   711.849976  

[8 rows x 50 columns]

In [5]:
print('List of the Stocks are : ')
df.columns[1:]

List of the Stocks are : 


Index(['RELIANCE.NS', 'HDFCBANK.NS', 'ICICIBANK.NS', 'INFY.NS', 'TCS.NS',
       'KOTAKBANK.NS', 'HINDUNILVR.NS', 'ITC.NS', 'LT.NS', 'SBIN.NS',
       'BAJFINANCE.NS', 'BHARTIARTL.NS', 'HCLTECH.NS', 'ASIANPAINT.NS',
       'AXISBANK.NS', 'DMART.NS', 'MARUTI.NS', 'ULTRACEMCO.NS', 'HDFC.NS',
       'TITAN.NS', 'SUNPHARMA.NS', 'M&M.NS', 'NESTLEIND.NS', 'WIPRO.NS',
       'ADANIGREEN.NS', 'TATASTEEL.NS', 'JSWSTEEL.NS', 'POWERGRID.NS',
       'ONGC.NS', 'NTPC.NS', 'COALINDIA.NS', 'BPCL.NS', 'IOC.NS', 'TECHM.NS',
       'INDUSINDBK.NS', 'DIVISLAB.NS', 'GRASIM.NS', 'CIPLA.NS',
       'BAJAJFINSV.NS', 'TATAMOTORS.NS', 'HEROMOTOCO.NS', 'DRREDDY.NS',
       'SHREECEM.NS', 'BRITANNIA.NS', 'UPL.NS', 'EICHERMOT.NS', 'SBILIFE.NS',
       'ADANIPORTS.NS', 'BAJAJ-AUTO.NS', 'HINDALCO.NS'],
      dtype='object')

In [6]:
df.shape

(24, 51)

In [7]:
# Converting date column to datetime:
df['Date'] = pd.to_datetime(df['Date'])

In [8]:
df['Date'].dtypes

datetime64[ns, UTC+05:30]

In [9]:
# Checking null values
print(df.isnull().sum())

Date              0
RELIANCE.NS       0
HDFCBANK.NS       0
ICICIBANK.NS      0
INFY.NS           0
TCS.NS            0
KOTAKBANK.NS      0
HINDUNILVR.NS     0
ITC.NS            0
LT.NS             0
SBIN.NS           0
BAJFINANCE.NS     0
BHARTIARTL.NS     0
HCLTECH.NS        0
ASIANPAINT.NS     0
AXISBANK.NS       0
DMART.NS          0
MARUTI.NS         0
ULTRACEMCO.NS     0
HDFC.NS          24
TITAN.NS          0
SUNPHARMA.NS      0
M&M.NS            0
NESTLEIND.NS      0
WIPRO.NS          0
ADANIGREEN.NS     0
TATASTEEL.NS      0
JSWSTEEL.NS       0
POWERGRID.NS      0
ONGC.NS           0
NTPC.NS           0
COALINDIA.NS      0
BPCL.NS           0
IOC.NS            0
TECHM.NS          0
INDUSINDBK.NS     0
DIVISLAB.NS       0
GRASIM.NS         0
CIPLA.NS          0
BAJAJFINSV.NS     0
TATAMOTORS.NS     0
HEROMOTOCO.NS     0
DRREDDY.NS        0
SHREECEM.NS       0
BRITANNIA.NS      0
UPL.NS            0
EICHERMOT.NS      0
SBILIFE.NS        0
ADANIPORTS.NS     0
BAJAJ-AUTO.NS     0


There are 24 null values in the closing prices of HDFC. 

In [10]:
df.drop(columns=['HDFC.NS'],inplace=True)

#### Now, we will look at the stock price trends of all the companies in the data:

In [11]:
fig = go.Figure()

for company in df.columns[1:]:
    fig.add_trace(go.Scatter(x=df['Date'], y=df[company], mode='lines', name=company, opacity=0.5))

fig.update_layout(title='Stock price trend of all Indian companies',xaxis_title='Date',yaxis_title='Closing Price (INR)', xaxis=dict(tickangle=45),legend=dict(x=1.05,y=1,traceorder="normal",font=dict(size=10), orientation="v"), margin=dict(l=0,r=0,t=30,b=0), hovermode='x',template='plotly_white')
fig.show()

#### -> Let’s look at the companies with the highest risks for investing:

In [12]:
all_companies = df.columns[1:]
volatility_all_companies = df[all_companies].std()
volatility_all_companies.sort_values(ascending=False).head(10)

BAJAJ-AUTO.NS    659.810841
SHREECEM.NS      429.919834
BAJFINANCE.NS    306.658594
DIVISLAB.NS      247.674895
HEROMOTOCO.NS    247.092728
DRREDDY.NS       175.124908
ULTRACEMCO.NS    172.673053
DMART.NS         155.593701
BRITANNIA.NS     144.164343
MARUTI.NS        109.587342
dtype: float64

### ->Now look at the companies with the highest growth rate for investing:

In [13]:
growth_all_companies = df[all_companies].pct_change() * 100
average_growth_all_companies = growth_all_companies.mean()
average_growth_all_companies.sort_values(ascending=False).head(10)

BAJAJ-AUTO.NS    0.883421
BAJAJFINSV.NS    0.791730
BHARTIARTL.NS    0.735219
DIVISLAB.NS      0.634851
HEROMOTOCO.NS    0.602192
ICICIBANK.NS     0.557742
BAJFINANCE.NS    0.536819
TITAN.NS         0.393800
HINDUNILVR.NS    0.351634
BRITANNIA.NS     0.327747
dtype: float64

### -> Now, let’s have a look at the companies with the highest return on investments:

In [14]:
initial_prices_all = df[all_companies].iloc[0]
final_prices_all = df[all_companies].iloc[-1]

roi_all_companies = ((final_prices_all - initial_prices_all) / initial_prices_all) * 100

roi_all_companies.sort_values(ascending=False).head(10)

BAJAJ-AUTO.NS    22.107017
BAJAJFINSV.NS    19.642973
BHARTIARTL.NS    18.120965
DIVISLAB.NS      15.404976
HEROMOTOCO.NS    14.660402
ICICIBANK.NS     13.480860
BAJFINANCE.NS    12.797149
TITAN.NS          9.275089
HINDUNILVR.NS     8.235039
BRITANNIA.NS      7.713587
dtype: float64

# Creating a Mutual Fund Plan Based on High ROI and Low Risk :

#### To create a strategy for selecting companies with high ROI and low risk, we can use a combination of ROI and volatility (standard deviation) metrics. 

#### The goal is to find companies that offer a high return on investment (ROI) but with low volatility to minimize risk.

##### the steps we can follow for creating a mutual fund plan:

1. Define ROI and Volatility Thresholds: We will set thresholds for ROI and volatility to select companies that provide good returns with lower risks.
2. Rank Companies by ROI and Volatility: Rank all companies based on their ROI and volatility scores.
3. Assign Investment Ratios: Allocate more investment to companies with higher ROI and lower volatility.


### Let’s start by defining thresholds and selecting companies that meet the criteria of high ROI and low volatility:

In [15]:
roi_threshold = roi_all_companies.median()
volatility_threshold = volatility_all_companies.median()

seleceted_companies = roi_all_companies[(roi_all_companies > roi_threshold) & (volatility_all_companies < volatility_threshold)]

seleceted_companies.sort_values(ascending=False)

ICICIBANK.NS     13.480860
INDUSINDBK.NS     7.159914
JSWSTEEL.NS       7.021748
AXISBANK.NS       6.592466
HDFCBANK.NS       6.319839
SUNPHARMA.NS      5.627425
KOTAKBANK.NS      5.474481
CIPLA.NS          4.850117
NTPC.NS           4.356926
dtype: float64

#### The following companies meet the criteria of high ROI and low volatility:

1. ICICI Bank (ROI: 13.48%)
2. IndusInd Bank (ROI: 7.16%)
3. JSW Steel (ROI: 7.02%)
4. Axis Bank (ROI: 6.59%)
5. HDFC Bank (ROI: 6.32%)
6. Sun Pharma (ROI: 5.63%)
7. Kotak Bank (ROI: 5.47%)
8. Cipla (ROI: 4.85%)
9. NTPC (ROI: 4.36%)


#### To balance the investment between these companies, we can use an inverse volatility ratio for allocation. Companies with lower volatility will get a higher weight. Let’s calculate the weight for each company:


In [16]:
selected_volatility = volatility_all_companies[seleceted_companies.index]
inverse_volatility = 1 / selected_volatility

inverse_ratios = inverse_volatility / inverse_volatility.sum()

inverse_ratios.sort_values(ascending=False)

NTPC.NS          0.280768
JSWSTEEL.NS      0.159985
AXISBANK.NS      0.092231
HDFCBANK.NS      0.089330
CIPLA.NS         0.084783
KOTAKBANK.NS     0.076642
INDUSINDBK.NS    0.074432
SUNPHARMA.NS     0.072553
ICICIBANK.NS     0.069276
dtype: float64

#### The investment ratios based on inverse volatility are as follows:

1. NTPC: 28.08%
2. JSW Steel: 15.99%
3. Axis Bank: 9.22%
4. HDFC Bank: 8.93%
5. Cipla: 8.48%
6. Kotak Bank: 7.66%
7. IndusInd Bank: 7.44%
8. Sun Pharma: 7.26%
9. ICICI Bank: 6.93%


## Analyzing Our Mutual Fund Plan

#### We have created a mutual fund plan for long-term investments. Now, let’s analyze and compare our mutual fund plan by comparing it with the high-performing companies in the stock market. Let’s start by comparing the risks in our mutual fund with the risk in the high growth companies:

In [17]:
top_growth_companies = average_growth_all_companies.sort_values(ascending=False).head(10)
risk_growth_rate_companies = volatility_all_companies[top_growth_companies.index]
risk_mutual_fund_companies = volatility_all_companies[seleceted_companies.index]

fig = go.Figure()

fig.add_trace(go.Bar(y=risk_mutual_fund_companies.index, x=risk_mutual_fund_companies, orientation='h', name='Mututal Fund Companies', marker = dict(color='blue')))

fig.add_trace(go.Bar(y=risk_growth_rate_companies.index, x=risk_growth_rate_companies, orientation='h', name='Growth Rate Companies', marker=dict(color='green'), opacity=0.7))

fig.update_layout(title='Risk Comparison: Mutual Fund vs Growth Rate Companies', xaxis_title='Volatility (Standard Deviation)', yaxis_title='Companies', barmode='overlay', legend=dict(title='Company Type'), template='plotly_white')

fig.show()

#### Now, let’s compare the ROI of both the groups as well:

In [18]:
expected_roi_mutual_fund = roi_all_companies[seleceted_companies.index]

expected_roi_growth_companies = roi_all_companies[top_growth_companies.index]

fig = go.Figure()

fig.add_trace(go.Bar(y=expected_roi_mutual_fund.index, x=expected_roi_mutual_fund, orientation='h', name='Mutual Fund Companies', marker=dict(color='blue')))

fig.add_trace(go.Bar(y=expected_roi_growth_companies.index, x=expected_roi_growth_companies, orientation='h', name='Growth Rate Companies', marker=dict(color='green'), opacity=0.7))

fig.update_layout(title='Expected ROI Comparison: Mutual Fund vs Growth Rate Companies', xaxis_title='Expected ROI (%)', yaxis_title='Companies', barmode='overlay',  legend=dict(title='Company Type'), template='plotly_white')

fig.show()

#### The comparison between the risk (volatility) and expected ROI for mutual fund companies (in blue) and growth rate companies (in green) shows a clear trade-off. Mutual fund companies offer lower volatility, meaning they are less risky, but also provide lower expected returns. In contrast, growth rate companies demonstrate higher volatility, indicating more risk, but they offer much higher potential returns, especially companies like Bajaj Auto and Bajaj Finserv. This highlights a common investment dilemma: lower risk comes with a lower reward, while higher risk could yield higher returns.

#### For long-term investments, the goal is typically to find companies that offer a balance of stable returns and manageable risk. The companies in our mutual fund exhibit low volatility, meaning they are less risky, and their moderate returns make them solid choices for long-term, stable growth. They are well-suited for conservative investors who want steady returns without significant fluctuations in value.

## Calculating Expected Returns

Now, let’s calculate the expected returns a person will get from our mutual fund if he/she invests ₹5000 every month.

To calculate the expected value a person will accumulate over 1 year, 3 years, 5 years, and 10 years through the mutual fund plan, we can follow these steps:

1. Assume the person is investing 5000 rupees every month.
2. Use the expected ROI from the mutual fund companies to simulate the growth over time.
3. Compute the compounded value of the investments for each period (1y, 3y, 5y, and 10y).
4. Visualize the accumulated value over these periods.


In [19]:
monthly_investment = 5000  # Monthly investment in INR
years = [1, 3, 5, 10]  # Investment periods (in years)
n = 12  # Number of times interest is compounded per year (monthly)

avg_roi = expected_roi_mutual_fund.mean() / 100  # Convert to decimal

def future_value(P, r, n, t):
    return P * (((1 + r/n)**(n*t) - 1) / (r/n)) * (1 + r/n)

future_values = [future_value(monthly_investment, avg_roi, n, t) for t in years]

fig = go.Figure()

fig.add_trace(go.Scatter(
    x=[str(year) + " year" for year in years],
    y=future_values,
    mode='lines+markers',
    line=dict(color='blue'),
    marker=dict(size=8),
    name='Future Value'
))

fig.update_layout(
    title="Expected Value of Investments of ₹ 5000 Per Month (Mutual Funds)",
    xaxis_title="Investment Period",
    yaxis_title="Future Value (INR)",
    xaxis=dict(showgrid=True, gridcolor='lightgrey'),
    yaxis=dict(showgrid=True, gridcolor='lightgrey'),
    template="plotly_white",
    hovermode='x'
)

fig.show()

#### After 1 year, the accumulated value is around ₹62,000, and by 5 years, it grows to over ₹300,000. The long-term benefit is evident, with the investment growing to nearly ₹860,000 over 10 years, which emphasises the value of consistent investing and compounding over time for long-term investors.